### ISSUE TITLE
Can one make SBT use custom scala-reflect.jar?

### ISSUE THREAD

**Issue by Blaisorblade** [sbt/sbt#626]
While scala-library and scala-compiler are hardcoded to be looked up in scalaHome, that does not extend to scala-reflect, and it's not clear how that works with other scala-provided JARs. This will become a more pressing problem as more and more of the library is modularized.

It seems that `scala-library` and `scala-compiler` are hardcoded to use a different lookup mechanism, which considers `scalaHome`, while other JARs aren't. Looking at the source, there are some provisions for other JARs, but they follow a different code path. E.g.:

``` scala
final class ScalaInstance(val version: String, val loader: ClassLoader, val libraryJar: File, val compilerJar: File, val extraJars: Seq[File], val explicitActual: Option[String]) extends xsbti.compile.ScalaInstance
```

EDIT: Here are two better instances, where scala artifacts name are hardcoded instead of discovered:

https://github.com/harrah/xsbt/blob/0.13/ivy/src/main/scala/sbt/IvyScala.scala#L24
https://github.com/harrah/xsbt/blob/0.13/ivy/src/main/scala/sbt/IvyScala.scala#L104

It seems that one should inspect `scalaHome / "lib"` to figure out which artifacts are available.

I take the description of symptoms from Paul Phillips's mail (https://groups.google.com/d/msg/scala-internals/v9mMzMBNlLI/tUvFj-2xRqIJ):

> I imagine what I was complaining about is the amazing difficulty I had getting sbt to use the scala-reflect.jar which went with the library and compiler jars, which I never did figure out how to make work.
> 
> ```
> // demonstration build.sbt
> name := "proj"
> 
> scalaVersion := "2.10.0-RC5"
> 
> scalaHome := Some(file("/scala/inst/scala-2.10.1-20121210-141937-7fe7d25379"))
> 
> logLevel in Global := Level.Debug
> 
> libraryDependencies <++= (scalaVersion)(sv =>
>   Seq(
>     "org.scala-lang" % "scala-reflect" % sv,
>     "org.scala-lang" % "scala-compiler" % sv
>   )
> )
> ```
> 
> If you use this to compile something, the debug output shows the classpath:
> 
> ```
> // [debug] Calling Scala compiler with arguments  (CompilerInterface):
> // [debug]   -d
> // [debug]   /project/Dec14-06PM/target/scala-2.10/classes
> // [debug]   -bootclasspath
> // [debug]   <snip jdk classes>:/scala/inst/scala-2.10.1-20121210-141937-7fe7d25379/lib/scala-library.jar
> // [debug]   -classpath <see below>
> 
> // Placed on individual lines for readability
> /project/Dec14-06PM/target/scala-2.10/classes
> /Users/paulp/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.10.0-RC5.jar
> /scala/inst/scala-2.10.1-20121210-141937-7fe7d25379/lib/scala-compiler.jar
> ```
> 
> So it's custom library, custom compiler, and stock reflect jar, despite my handling reflect and compiler identically and also despite the fact that scala-compiler depends on scala-reflect (in real life: I don't know if that's true in the mind of sbt.) Setting scalaVersion to that exact version only leads to it failing to find any scala-reflect.jar.

The funny thing is that when compiling compiler-interface, the classpath is setup correctly:

```
[info] 'compiler-interface' not yet compiled for Scala 2.10.0-RC2. Compiling...
[debug] Plain interface to Scala compiler 2.10.0-RC2  with arguments: 
[debug]     -nowarn
[debug]     -d
[debug]     /var/folders/_7/hlxv4yv95x95vgnn416b4q4m0000gn/T/sbt_30c20d7
[debug]     -bootclasspath
[debug]     /System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/jsfd.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/classes.jar:/System/Library/Frameworks/JavaVM.framework/Frameworks/JavaRuntimeSupport.framework/Resources/Java/JavaRuntimeSupport.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/ui.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/laf.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/sunrsasign.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/jsse.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/jce.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/charsets.jar:/Users/pgiarrusso/Documents/Research/Sorgenti/scala-2.10.0-RC2/lib/scala-library.jar
[debug]     -classpath
[debug]     /Users/pgiarrusso/.sbt/boot/scala-2.9.2/org.scala-sbt/sbt/0.12.1/xsbti/interface-0.12.1.jar:/Users/pgiarrusso/.sbt/boot/scala-2.9.2/org.scala-sbt/sbt/0.12.1/compiler-interface-src/compiler-interface-src-0.12.1.jar:/Users/pgiarrusso/Documents/Research/Sorgenti/scala-2.10.0-RC2/lib/scala-compiler.jar:/Users/pgiarrusso/Documents/Research/Sorgenti/scala-2.10.0-RC2/lib/jline.jar:/Users/pgiarrusso/Documents/Research/Sorgenti/scala-2.10.0-RC2/lib/scala-reflect.jar
[debug]     /var/folders/_7/hlxv4yv95x95vgnn416b4q4m0000gn/T/sbt_eb5a0caf/API.scala
```

but not doing real compilation:

```
[debug] Calling Scala compiler with arguments  (CompilerInterface):
[debug]     -d
[debug]     /Users/pgiarrusso/Documents/Research/Sorgenti-Sync/Scala/importSpecificity/target/scala-2.10/classes
[debug]     -bootclasspath
[debug]     /System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/jsfd.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/classes.jar:/System/Library/Frameworks/JavaVM.framework/Frameworks/JavaRuntimeSupport.framework/Resources/Java/JavaRuntimeSupport.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/ui.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/laf.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/sunrsasign.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/jsse.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/jce.jar:/System/Library/Java/JavaVirtualMachines/1.6.0.jdk/Contents/Classes/charsets.jar:/Users/pgiarrusso/Documents/Research/Sorgenti/scala-2.10.0-RC2/lib/scala-library.jar
[debug]     -classpath
[debug]     /Users/pgiarrusso/Documents/Research/Sorgenti-Sync/Scala/importSpecificity/target/scala-2.10/classes:/Users/pgiarrusso/.ivy2/cache/org.scala-lang/scala-reflect/jars/scala-reflect-2.10.0-RC2.jar:/Users/pgiarrusso/Documents/Research/Sorgenti/scala-2.10.0-RC2/lib/scala-compiler.jar
```

**Comment by harrah** [comment]
The linked examples of hardcoding don't apply to the situation here.  One is for excluding Scala libraries, which is no longer used and the underlying code may be deprecated at some point.  The other is for pulling in the compiler so that scalac and scaladoc can be invoked and it isn't used when scalaHome is set.  Although it would be possible to allow the user to specify the module names for the library and compiler jars, it wouldn't affect anything here.

The problem is mixing managed and unmanaged dependencies.  There was no intention previously to allow mixing managed dependencies on Scala jars and have those translated to use the jars in Scala home.  (Although scala-library.jar and scala-compiler.jar get translated, it is for another reason.)  It is always possible to use custom Scala jars by publishing them to a repository with a custom version.  The issues come up when trying to mix scalaHome with managed dependencies.

There are improvements to Scala handling in 0.13, but in general, I expect it to be tricky to get the edge cases between managed/unmanaged Scala dependencies right.  I can see that it can be useful to take an existing build and simply set scalaHome to use a local Scala version, but it is still tricky to define unsurprising semantics.  I've written up the expected behavior in the 0.13 documentation here:

 https://github.com/harrah/xsbt/blob/0.13/src/sphinx/Detailed-Topics/Configuring-Scala.rst

This makes it an explicit goal to now have scalaHome provide the jars for managed dependencies `"org.scala-lang" % X % "<version>"` from `<scala-home>/lib/X.jar` when it exists.  It is necessarily approximation subject to ongoing refinement.  For example, the Scala home directory includes no dependency information (and probably it would be hard to integrate it if there were such information), so the dependencies are being approximated by the resolved metadata for whatever scalaVersion is set for the project.  If you set scalaVersion to 2.9.2 but compile 2.10.Y from source, you won't get your scala-reflect.jar as a dependency of scala-compiler.jar.

The last bits related to this unmanaged/managed overlap were implemented in d156ccfe4efec7ea249869a5a493f1fbe773da58.

**Comment by harrah** [comment]
It seems the primary resolution wasn't communicated well, sorry.
1. Scala handling in general was already improved in 0.13.
2. This bug was accepted as the triple Bug+Enhancement+Documentation and was addressed in d156ccfe4efec7ea249869a5a493f1fbe773da58.
3. The expected behavior is now documented on the [Configuring Scala](https://github.com/harrah/xsbt/blob/0.13/src/sphinx/Detailed-Topics/Configuring-Scala.rst) page.  Please take a look and try it out once a 0.13 milestone is out.  The interaction between managed+unmanaged is tricky, so it will probably take more iterations to get it right.
